### PR TITLE
Task/reject call actions

### LINKED ISSUE #167: Add reject pending intent and action to PN integration
As seen [in the push PR](https://github.com/GetStream/stream-video-android/pull/137/files#diff-78fedeb6e908233ffd65b8539f533f0bf19b251452cc13b18ffde9516a23958aR117), the action for rejecting/declining a call is currently empty (an empty intent).

We should build an intent similar to acceptPendingIntent that has the DECLINE action and is hooked up to our abstract defaults to reject the call and finish itself.

### Context & proposal

Build a function that returns a pending intent to decline a call and plug it into the decline action.

### Acceptance criteria

To accept this task we need to complete the following items:

- [x] Define a decline call action in push.
- [x] Build a pending intent to decline a call.
- [x] Update the PN guide to include this decline/reject logic.

### PR DESCRIPTION
**Pull request by filbabic**
### 🎯 Goal

Closes #167 by adding necessary actions and internal logic to reject a call through push.

### 🛠 Implementation details

Added a broadcast receiver that consumes pending intents to reject a call. Added logic to choose between the Receiver and an Activity based on the Action.

### 🧪 Testing

1. Run the Dogfooding app on an emulator.
2. Make sure you are logged in (either your Stream Google acc or random guest sign up)
3. Test several scenarios - App in FG, App in BG, App Killed.
4. Follow the patch note commands to trigger an ADB data notification.
5. Once the notification is shown, press the "Reject" action. It should dismiss the notification and reject the call.
6. Observe the logs: [onReceive] in Logcat and its result. It should show successful result.

<details>

<summary>Provide the patch summary here</summary>

```
adb root


adb shell am broadcast -a com.google.android.c2dm.intent.RECEIVE -n io.getstream.video.android.dogfooding.debug/com.google.firebase.iid.FirebaseInstanceIdReceiver -a "com.google.android.c2dm.intent.RECEIVE" --es "sender" "stream.video" --es "type" "call_incoming" --es "call_cid" "default:call_123" --es "user_names" "Jc\,\ Filip"
```

</details>

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (required internally)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] Compose sample runs & works
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+                CoroutineScope(Dispatchers.IO).launch {
+                    when (val rejectResult = streamVideo.rejectCall(callCid)) {
+                        is Success -> logger.d { "[onReceive] rejectCall, Success: $rejectResult" }
+                        is Failure -> logger.d { "[onReceive] rejectCall, Failure: $rejectResult" }
+                    }
+
+                    val notificationId = intent.getIntExtra(INTENT_EXTRA_NOTIFICATION_ID, 0)
+                    NotificationManagerCompat.from(context).cancel(notificationId)
+                }
```
Comment (JcMinarro) [inline stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt]: Is it fast enough that we don't need to start a foreground service? :thinking:

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt [hunk 2]
```diff
@@ -56,12 +57,8 @@ internal class VideoPushDelegate(
...
                 it.createNotificationChannel(
                     NotificationChannelCompat.Builder(
-                        CHANNEL_ID,
-                        NotificationManager.IMPORTANCE_HIGH
-                    )
-                        .setName(CHANNEL_NAME)
-                        .setDescription(CHANNEL_DESCRIPTION)
-                        .build()
+                        CHANNEL_ID, NotificationManager.IMPORTANCE_HIGH
+                    ).setName(CHANNEL_NAME).setDescription(CHANNEL_DESCRIPTION).build()
```
Comment (JcMinarro) [inline stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt]: Why this change?
Isn't it better one method per line to avoid merge conflicts?

File: stream-video-android/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -20,19 +22,25 @@
...
-    <application >
+    <application>
 
         <provider
             android:name="io.getstream.android.push.delegate.PushDelegateProvider"
             android:authorities="${applicationId}.io.getstream.android.push"
             android:exported="false"
-            tools:node="merge"
-            >
+            tools:node="merge">
```
Comment (JcMinarro) [inline stream-video-android/src/main/AndroidManifest.xml]: Leaving a new line before the ending tag helps us in the future with merge conflicts

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt [hunk 4]
```diff
@@ -78,56 +75,44 @@ internal class VideoPushDelegate(
...
     ) {
         val notification = NotificationCompat.Builder(context, CHANNEL_ID)
             .setSmallIcon(androidx.loader.R.drawable.notification_bg)
-            .setContentTitle("Incoming call")
-            .setContentText(users)
-            .setOngoing(false)
-            .setAutoCancel(true)
-            .setContentIntent(fullScreenPendingIntent)
+            .setContentTitle("Incoming call").setContentText(users).setOngoing(false)
+            .setAutoCancel(true).setContentIntent(fullScreenPendingIntent)
```
Comment (JcMinarro) [inline stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt]: Same here

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt [hunk 5]
```diff
@@ -146,40 +131,116 @@ internal class VideoPushDelegate(
...
-                        putExtra(INTENT_EXTRA_NOTIFICATION_ID, INCOMING_CALL_NOTIFICATION_ID)
-                    },
-                    flags
-                )
+
+                if (isRejectCall) {
+                    getBroadcastForIntent(baseIntent, resolveInfo, callId, flags)
+                } else {
+                    getActivityForIntent(baseIntent, resolveInfo, callId, flags)
+                }
```
Comment (JcMinarro) [inline stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt]: Can we split it into two methods and reuse the common logic?
Something like `searchActivityPendingIntent()` and `searchBroadcastPendingIntent()`

File: stream-video-android/src/main/AndroidManifest.xml (hunk 3 above)
Comment (filbabic) [reply stream-video-android/src/main/AndroidManifest.xml]: Good point - auto linter just did this

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt (hunk 4 above)
Comment (filbabic) [reply stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt]: I think the linting/formatting did this, definitely makes sense to make it all a separate line

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt (hunk 5 above)
Comment (filbabic) [reply stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt]: Hmmm but we are already doing that?

getBroadcastForIntent and getActivityForIntent reuses common logic to build the component info and just replaces if it's going for a broadcast or an activity

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt (hunk 1 above)
Comment (filbabic) [reply stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt]: Hmmm fair enough. In the case that I tested, the notification was just present until this API call finished, so it's there for ~1 sec before dismissing.

Do you have any proposals as to what we could do in this case or how to stress test it?

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt (hunk 2 above)
Comment (filbabic) [reply stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/VideoPushDelegate.kt]: Auto format I think, will fix

File: stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt (hunk 1 above)
Comment (JcMinarro) [reply stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt]: Android OS will kill the Broadcast Receiver if it expends more than 5 seconds.
It shouldn't be the case and our `StreamVideo` should respond before that.
You should be able to do some tests by adding some `Thread.sleep()`  on this code and check what happens

Comment (JcMinarro) [inline stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt]:
```suggestion
                CoroutineScope(Dispatchers.IO).launch {
                    when (val rejectResult = streamVideo.rejectCall(callCid)) {
                        is Success -> logger.d { "[onReceive] rejectCall, Success: $rejectResult" }
                        is Failure -> logger.d { "[onReceive] rejectCall, Failure: $rejectResult" }
                    }
                }
                val notificationId = intent.getIntExtra(INTENT_EXTRA_NOTIFICATION_ID, 0)
                NotificationManagerCompat.from(context).cancel(notificationId)
```
BTW, It is a "fire&forget" call, we can call it and dismiss the notification without wait until our `StreamVideo` responds

Comment (filbabic) [reply stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt]: Yeah that sounds good enough, in case of 5 seconds, the API is in serious trouble anyway, so the call might just time out on its own

Comment (filbabic) [reply stream-video-android/src/main/kotlin/io/getstream/video/android/notifications/internal/RejectCallBroadcastReceiver.kt]: Fair enough! I figured we wouldn't remove the notif until it's confirmed, but at the same time, if the user rejected, might as well remove the notif immediately
